design/036-package_id.md

```diff
@@ -0,0 +1,107 @@
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [036](https://github.com/conan-io/tribe/pull/36)  |
+| **Submitted**     | 2022-06-17                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+This proposal is about the effect that the dependencies have over the current package ``package_id``, but not about other factors, such as the package settings, options, configuration or other type of binary compatibility resulting from other factors other than dependencies.
+
+Conan 2.0 proposes a new default ``package_id`` that will be dynamic, based on two main cases:
+
+- When an application is linking a static library, or a shared library is linking a static library, or any library or application is linking a header-only library as an implementation detail, the resulting artifacts will be fully embedding/inlining the dependency artifacts, and they require in the general case to be re-built from sources when something in the dependency changes. This will be the "embed" case, and by default will use a "full_mode" policy, which means that the full dependency reference ``pkg/version@user/channnel#recipe_revision:package_id`` will be used and factored in the consumer ``package_id``. Recall that ``package_revision`` can no longer be part of the ``package_id`` as removed in proposal https://github.com/conan-io/tribe/pull/30
+- When a static library is depending on a static library, or a header-only library is depending on another library, the resulting artifacts in general do not embed/inline the dependency artifacts, and they don’t need to be rebuilt in many cases even if the dependency changes. For example, doing a patch in an implementation file .cpp of a static library does not require another static library that depends on it to be recompiled. This will be the "non embed" case and by default will use a "minor_mode" policy, which means requiring a rebuild if the major or minor of the dependency changes, but not requiring it if the patch of the dependency changes.
+
+Conan will be using the ``package_type`` information to define which of the above scenarios apply.
+
+Conan 2.0 will allow ``tool_requires`` (or any other ``requires`` with trait ``build=True``) to affect the ``package_id`` of their consumers. By default it will be effect will be ``None``, but users will be able to define it, globally via ``global.conf`` configuration, or per recipe in the ``requires(...., build=True, package_id_mode=xxxx)`` trait (being ``xxx`` any valid mode like ``minor_mode``, ``patch_mode``, ``full_mode``, etc)
+
+Conan 2.0 ``package_id`` will not depend by default on the dependencies’ ``options``. Their influence should be represented in most cases by the scenarios above, and the ``full_mode`` taking into account the dependency ``package_id`` into the consumer ``package_id``.
```

Comment:
I think about one case (which can be presented as an example) - when option affects compile-time definitions which result in different ABI. For example (I may be wrong, but still), in ASIO one can chose between `io_uring` implementation or `epoll_reator` which in the end will provide different type aliases and therefore making consumer incompatible with one built with different options (even if consumer is a static library).

I think another case, when options affect dependencies like `with_xxx`, is covered indirectly by always requiring whole graph, so flipping such flag will produce another `package_id`, but because the graph has changed.

(Now I started wondering about whole boost or openssl...)

Otherwise I think not requiring options by default is a good change - I think I filed 2 or 3 issues myself when options resulted in some issues.